tensorboard/java/org/tensorflow/tensorboard/vulcanize/Vulcanize.java

```diff
@@ -793,22 +802,73 @@ private static void combineScriptElements(Document document) {
       hashes.add(hash);
     }
     return hashes;
   }
 
-  // Writes sha256 of script tags in base64 in the document.
-  private static void writeShasum(Document document, Path output) throws FileNotFoundException, IOException {
-    String hashes = Joiner.on("\n").join(computeScriptShasum(document));
-    Files.write(
-        output,
-        hashes.getBytes(UTF_8),
-        StandardOpenOption.WRITE,
-        StandardOpenOption.CREATE,
-        StandardOpenOption.TRUNCATE_EXISTING);
+  /**
+   * Writes sha256 of script tags in base64 in the document.
+   *
+   * @deprecated Shasum is deprecated in favor of script extraction.
+   */
+  @Deprecated
+  private static String getScriptsShasums(Document document)
+      throws FileNotFoundException, IOException {
+    return Joiner.on("\n").join(computeScriptShasum(document));
+  }
+
+  private static String extractScriptContent(Document document)
+      throws FileNotFoundException, IOException, IllegalArgumentException {
+    Elements scripts = document.getElementsByTag("script");
+    StringBuilder sourcesBuilder = new StringBuilder();
+
+    for (Element script : scripts) {
+      String sourceContent;
+      String src = script.attr("src");
+      if (src.isEmpty()) {
+        sourceContent = script.html();
+      } else {
+        // script element that remains are the ones with src that is absolute or annotated with
+        // `jscomp-ignore`. They must resolve from the root because those srcs are rootified.
+        Webpath webpathSrc = Webpath.get(src);
+        Webpath webpath = Webpath.get("/").resolve(webpathSrc).normalize();
+        if (isAbsolutePath(webpathSrc)) {
+          if (script.hasAttr("defer") || script.hasAttr("async")) {
+            continue;
+          }
+          throw new IllegalArgumentException(
+              "Script refers to a remote resource ("
+                  + webpathSrc
+                  + ") in a blocking way. For"
+                  + " correctness of execution, please make sure it is async-able or defer-able:"
+                  + script.outerHtml());
+        } else if (!webfiles.containsKey(webpath)) {
+          throw new FileNotFoundException(
+              "Expected webfiles for " + webpath + " to exist. Related: " + script.outerHtml());
+        }
+        sourceContent = new String(Files.readAllBytes(webfiles.get(webpath)), UTF_8);
+      }
+
+      sourcesBuilder.append(sourceContent).append("\n");
+      script.remove();
+    }
+
+    return sourcesBuilder.toString();
+  }
+
+  private static String extractAndTransformJavaScript(Document document, Webpath scriptPath)
+      throws FileNotFoundException, IOException, IllegalArgumentException {
+    String scriptContent = extractScriptContent(document);
+
+    Element lastBody = Iterables.getLast(document.getElementsByTag("body"));
+    Element scriptTag = new Element(Tag.valueOf("script"), "");
```

Comment:
Half-nit, but calling an element a tag is a bit confusing, especially
when the rest of the code clearly and correctly distinguishes them;
consider `scriptElement` or just `script`?

Reply:
Ah, sorry, didn’t realize that there actually was precedent in this
file. Either’s fine then; thanks.